share/sysbuild/cmake/modules/sysbuild_extensions.cmake

```
@@ -380,5 +380,58 @@ function(sysbuild_module_call)
       endforeach()
     endif()
   endforeach()
+endfunction()
+
+# Usage:
+#   sysbuild_cache_set(VAR <variable> [APPEND [LIST]] <value>)
+#
+# This function will set the specified value of the sysbuild cache variable in
+# the CMakeCache.txt file which can then be accessed by images.
+# `VAR` specifies the variable name to set/update.
+#
+# The result will be returned in `<variable>`.
+#
+# Example use:
+#   sysbuild_cache_set(VAR ATTRIBUTES APPEND LIST battery)
+#     Will add the item `battery` to the `ATTRIBUTES` variable as a new element
+#     in the list in the CMakeCache.
+#
+# <variable>: Name of variable in CMake cache.
+# APPEND:     If specified then will append the supplied data to the existing
+#             value (unless LIST is also specified, this will be treated as a
+#             string, no space character is added between appends).
+# LIST:       Will treat the append item as a list and add the new data as a
+#             new element to the list.
+# <value>:    Value to set/update.
+function(sysbuild_cache_set)
+  cmake_parse_arguments(VARS "APPEND;LIST" "VAR" "" ${ARGN})
+
+  if(NOT VARS_UNPARSED_ARGUMENTS AND VARS_APPEND)
+    message(FATAL_ERROR "sysbuild_set(VAR <var> APPEND ...) missing required arguments to append")
+  elseif(VARS_LIST AND NOT VARS_APPEND)
+    message(FATAL_ERROR "sysbuild_set(VAR <var> APPEND LIST ...) is required for the LIST option")
+  elseif(NOT VARS_VAR)
+    message(FATAL_ERROR "sysbuild_set(VAR <var> ...) missing required VAR variable")
+  endif()
+
+  get_property(var_type CACHE ${VARS_VAR} PROPERTY TYPE)
+
+  # If the variable type is not set, use UNINITIALIZED which will not apply any
+  # specific formatting.
+  if(NOT var_type)
+    set(var_type "UNINITIALIZED")
+  endif()
+
+  if(VARS_APPEND)
+    set(var_new $CACHE{${VARS_VAR}})
+    if(VARS_LIST)
+      list(APPEND var_new ${VARS_UNPARSED_ARGUMENTS})
```

I assume that this results in an ever expanding list in the CMake cache.
Maybe the value should only be appended if it doesn't exists in the cache list already.

Something like:
```suggestion
      foreach(new_val ${VARS_UNPARSED_ARGUMENTS})
        if(NOT new_val IN_LIST var_new)
          list(APPEND var_new ${new_val})
        endif()
      endforeach()
```
Note: code snippet above has not been validated.

---

Have added an option to remove duplicates but also allow to keep them e.g. someone might be construct a command line which uses the same argument multiple times for different values

---

>  e.g. someone might be construct a command line which uses the same argument multiple times for different values

makes a lot of sense.
Although when working with cache values and CMake re-runs, this is one of the tricky parts to get right, and we want to avoid ever growing list because of CMake re-runs.

CMake lists are effectively `;`-separated strings, so an option can be to search the list for pairs / n-length lists before appending, for example using: `string(FIND ...)`.

So you could search the list for such pairs, and if already existing, then not adding the duplicate.

As example:
```
set(list_var --foo)
set(bar1 --bar once)
set(bar2 --bar twice ${bar2})

list(APPEND list_var ${bar1})
list(REMOVE_DUPLICATES list_var)                                                                                                                                                            

list(APPEND list_var ${bar2})
list(REMOVE_DUPLICATES list_var)                                                                                                                                                            

list(APPEND list_var ${bar1})
list(REMOVE_DUPLICATES list_var)                                                                                                                                                            

list(APPEND list_var ${bar2})
list(REMOVE_DUPLICATES list_var)                                                                                                                                                            
```

gives the following list `--foo;--bar;once;twice`, and not `--foo;--bar;once;--bar;twice`

But if using `string(FIND ...)` before adding a new set of arguments, similar to this:
```
string(FIND "${list_var}" "${bar1}" index)                                                                                                                                                    
if(${index} EQUAL -1)                                                                                                                                                                         
  list(APPEND list_var ${bar1})                                                                                                                                                               
endif()                                                                                                                                                                                       
                                                                                                                                                                                              
string(FIND "${list_var}" "${bar2}" index)                                                                                                                                                    
if(${index} EQUAL -1)                                                                                                                                                                         
  list(APPEND list_var ${bar2})                                                                                                                                                               
endif()                                                                                                                                                                                       
                                                                                                                                                                                              
string(FIND "${list_var}" "${bar1}" index)                                                                                                                                                    
if(${index} EQUAL -1)                                                                                                                                                                         
  list(APPEND list_var ${bar1})                                                                                                                                                               
endif()                       

string(FIND "${list_var}" "${bar2}" index)                                                                                                                                                    
if(${index} EQUAL -1)                                                                                                                                                                         
  list(APPEND list_var ${bar2})                                                                                                                                                               
endif()                                                                                                                                                                                       
```
then we avoid duplicated entries / ever growing list, while obtaining exactly the list we want, which is: `--foo;--bar;once;--bar;twice`

---

Note: the `REMOVE_DUPLICATES` flag still makes a lot of sense to ensure a list of unique items.

The addition proposed in the above comment is to avoid ever growing lists of identical options / sub-lists / strings to be added on CMake re-runs.

---

Have added this but I'm not particularly happy with how the functionality differs from list(APPEND ...) - not that I would envisage someone using the cmake cache for a command but someone might have a command added with verbosity levels which get increased, so "-v" might be added, then later on "-v" might get added, then finally another 2 "-v"'s get added, all done without removing duplicates, but with this code right now only the first "-v" would be added. But to counter that if you did a cmake rerun without the string check you would end up with "-v -v -v -v -v -v -v -v" on the rerun so is a challenge.

---

>  but I'm not particularly happy with how the functionality differs from list(APPEND ...) 

no, but there is probably a good reason why CMake doesn't support `CACHE` and `FORCE` together with the [list](https://cmake.org/cmake/help/latest/command/list.html) function.

This is hinting that we are doing something we probably shouldn't be doing in the first place with this feature 😉 .
However, as we need the functionality provided with this PR, then I believe it's better to introduce `sysbuild_cache_set()` and limit the functionality so we prevent what would normally be unwanted behavior.

Those that need an ever growing list can create one such list by using CMake's regular `set()` function.
Remember, this PR is merely just offering an improved CMake cache handling for sysbuild with a built-in safety net.

> but with this code right now only the first "-v" would be added. 

Actually, if you do proper quoting of variables in `sysbuild_cache_set()`, then it should be possible to do:
```
sysbuild_cache_set(VAR CACHE_VAR APPEND "-v")
sysbuild_cache_set(VAR CACHE_VAR APPEND "-v\;")
```
or alternatively:
```
sysbuild_cache_set(VAR CACHE_VAR APPEND "-v;-v")
```

although in the last alternative user would need to ensure a single `-v` har not been added already.